pkg/apis/pipeline/v1beta1/task_types.go

```diff
@@ -78,6 +79,61 @@ func (*Task) GetGroupVersionKind() schema.GroupVersionKind {
 	return SchemeGroupVersion.WithKind(pipeline.TaskControllerName)
 }
 
+func findWorkspaceSubstitutionLocationsInSidecars(sidecars []Sidecar) sets.String {
+	locationsToCheck := sets.NewString()
+	for _, sidecar := range sidecars {
+		locationsToCheck.Insert(sidecar.Script)
+
+		for i := range sidecar.Args {
+			locationsToCheck.Insert(sidecar.Args[i])
+		}
+
+		for i := range sidecar.Command {
+			locationsToCheck.Insert(sidecar.Command[i])
+		}
+	}
+	return locationsToCheck
+}
+
+func findWorkspaceSubstitutionLocationsInSteps(steps []Step) sets.String {
+	locationsToCheck := sets.NewString()
+	for _, step := range steps {
+		locationsToCheck.Insert(step.Script)
+
+		for i := range step.Args {
+			locationsToCheck.Insert(step.Args[i])
+		}
+
+		for i := range step.Command {
+			locationsToCheck.Insert(step.Command[i])
+		}
+	}
+	return locationsToCheck
+}
+
+func findWorkspaceSubstitutionLocationsInStepTemplate(stepTemplate *StepTemplate) sets.String {
+	locationsToCheck := sets.NewString()
+
+	if stepTemplate != nil {
+		for i := range stepTemplate.Args {
+			locationsToCheck.Insert(stepTemplate.Args[i])
+		}
+		for i := range stepTemplate.Command {
+			locationsToCheck.Insert(stepTemplate.Command[i])
+		}
+	}
+	return locationsToCheck
+}
+
+// FindWorkspaceSubstitutionLocationsInTask returns a set of all the locations in the TaskSpec  where we can apply substitutions.
+func (ts *TaskSpec) FindWorkspaceSubstitutionLocationsInTask() sets.String {
```

Comment:
Where is this function being used? It looks like this is functionality for replacing `workspaces.foo.path` syntax but I don't see it being used anywhere-- is this replacement happening as part of this PR?
If this function does need to be in this PR, I don't think it should be exported.

Reply:
Good point. Even though it is written in task_types, it might be being called from the PR handling propagated workspaces in pipelinerun.

Reply:
This function was being called in the pipelinerun reconciler to parse the `pipelineTask` and identify which workspaces are being called. The reconciler then propagated those workspaces to the pipelineTasks by creating task bindings. 

The taskRun reconciler (`reconciler/taskrun/taskrun.go`) does a similar thing and propagates workspaces to the taskSpec.
The substitutions happen the way they did before. i.e. via https://github.com/chitrangpatel/pipeline/blob/e302587a9dcbc7af8cc1475dfa24c3c35a6715f9/pkg/reconciler/pipelinerun/resources/apply.go#L173-L185.